File: blind75/57_same_tree/same_tree.go
package same_tree

// TreeNode represents a node in a binary tree.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// IsSameTree checks if two binary trees are identical.
// Uses recursive DFS for O(n) time complexity.
func IsSameTree(p *TreeNode, q *TreeNode) bool {
	// Both nodes are nil - same structure
	if p == nil && q == nil {
		return true
	}

	// One node is nil, the other isn't - different structure
	if p == nil || q == nil {
		return false
	}

	// Values differ - not the same
	if p.Val != q.Val {
		return false
	}

	// Check if left and right subtrees are the same
	return IsSameTree(p.Left, q.Left) && IsSameTree(p.Right, q.Right)
}

File: blind75/57_same_tree/same_tree.py
from typing import Optional


class TreeNode:
    """Definition for a binary tree node."""

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def is_same_tree(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
    """
    Checks if two binary trees are identical.
    Uses recursive DFS for O(n) time complexity.

    Args:
        p: Root node of first binary tree
        q: Root node of second binary tree

    Returns:
        True if trees are identical, False otherwise
    """
    # Both nodes are None - same structure
    if p is None and q is None:
        return True

    # One node is None, the other isn't - different structure
    if p is None or q is None:
        return False

    # Values differ - not the same
    if p.val != q.val:
        return False

    # Check if left and right subtrees are the same
    return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)
